tests/pytests/integration/runners/test_vault.py

```diff
@@ -0,0 +1,285 @@
+    }
+    factory = salt_factories.salt_master_daemon(
+        "vault-pillarpolicy-functional-master", defaults=config_defaults
+    )
+    with factory.started():
+        yield factory
+
+
+@pytest.fixture(scope="module")
+def pillar_salt_minion(pillar_salt_master):
+    assert pillar_salt_master.is_running()
+    factory = pillar_salt_master.salt_minion_daemon(
+        "vault-pillarpolicy-functional-minion-1",
+        defaults={"open_mode": True, "grains": {"foo": "bar"}},
+    )
+    with factory.started():
+        # Sync All
+        salt_call_cli = factory.salt_call_cli()
+        ret = salt_call_cli.run("saltutil.sync_all", _timeout=120)
+        assert ret.returncode == 0, ret
+        yield factory
+
+
+@pytest.fixture(scope="module")
+def pillar_salt_run_cli(pillar_salt_master):
+    return pillar_salt_master.salt_run_cli()
+
+
+@pytest.fixture(scope="module")
+def pillar_salt_call_cli(pillar_salt_minion):
+    return pillar_salt_minion.salt_call_cli()
+
+
+@pytest.fixture()
+def pillar_policy_tree(pillar_state_tree, pillar_salt_minion, pillar_salt_call_cli):
+    top_file = """
+    base:
+      '{}':
+        - roles
+        - sdb_loop
+    """.format(
+        pillar_salt_minion.id
+    )
+    roles_pillar = """
+    roles:
+      - minion
+      - web
+    """
+    sdb_loop_pillar = """
+    foo: {{ salt["sdb.get"]("sdb://sdbvault/does/not/matter/val") }}
+    """
+    top_tempfile = pytest.helpers.temp_file("top.sls", top_file, pillar_state_tree)
+    roles_tempfile = pytest.helpers.temp_file(
+        "roles.sls", roles_pillar, pillar_state_tree
+    )
+    sdb_loop_tempfile = pytest.helpers.temp_file(
+        "sdb_loop.sls", sdb_loop_pillar, pillar_state_tree
+    )
+
+    with top_tempfile, roles_tempfile, sdb_loop_tempfile:
+        yield
+
+
+@pytest.fixture()
+def pillar_loop(
+    pillar_state_tree, pillar_policy_tree, pillar_salt_minion, pillar_salt_call_cli
+):
+    top_file = """
+    base:
+      '{}':
+        - roles
+        - sdb_loop
+        - exe_loop
+    """.format(
+        pillar_salt_minion.id
+    )
+    exe_loop_pillar = r"""
+    bar: {{ salt["vault.read_secret"]("does/not/matter") }}
+    """
+    top_tempfile = pytest.helpers.temp_file("top.sls", top_file, pillar_state_tree)
+    exe_loop_tempfile = pytest.helpers.temp_file(
+        "exe_loop.sls", exe_loop_pillar, pillar_state_tree
+    )
+
+    with top_tempfile, exe_loop_tempfile:
+        yield
+
+
+class TestVaultPillarPolicyTemplatesWithoutCache:
+    @pytest.fixture()
+    def minion_data_cache_absent(
+        self, pillar_salt_run_cli, pillar_salt_minion, pillar_policy_tree
+    ):
+        ret = pillar_salt_run_cli.run(
+            "cache.flush", f"minions/{pillar_salt_minion.id}", "data"
+        )
+        assert ret.returncode == 0
+        cached = pillar_salt_run_cli.run(
+            "cache.fetch", f"minions/{pillar_salt_minion.id}", "data"
+        )
+        assert cached.returncode == 0
+        assert not cached.data
```

Comment:
This is where the tests tend to fail. The most reproducible solution would probably to use one master with "minion_data_cache=False" and the default one for the cached tests.